patches/api/0347-Add-PlayerItemFrameChangeEvent.patch

```diff
@@ -0,0 +1,105 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: SamB440 <sam@islandearth.net>
+Date: Mon, 15 Nov 2021 18:09:46 +0000
+Subject: [PATCH] Add PlayerItemFrameChangeEvent
+
+
+diff --git a/src/main/java/io/papermc/paper/event/player/PlayerItemFrameChangeEvent.java b/src/main/java/io/papermc/paper/event/player/PlayerItemFrameChangeEvent.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..425cab00fe162588706c91385a5aa3d2de8f47f5
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/event/player/PlayerItemFrameChangeEvent.java
+@@ -0,0 +1,93 @@
++package io.papermc.paper.event.player;
++
++import org.bukkit.entity.ItemFrame;
++import org.bukkit.entity.Player;
++import org.bukkit.event.Cancellable;
++import org.bukkit.event.HandlerList;
++import org.bukkit.event.player.PlayerEvent;
++import org.bukkit.inventory.ItemStack;
++import org.jetbrains.annotations.NotNull;
++
++/**
++ * Called when an {@link ItemFrame} is having an item rotated, added, or removed from it.
++ */
++public class PlayerItemFrameChangeEvent extends PlayerEvent implements Cancellable {
++
++    private static final HandlerList HANDLERS = new HandlerList();
++    private boolean cancelled;
++    private final ItemFrame itemFrame;
++    private ItemStack itemStack;
++    private final ItemFrameChangeAction action;
++
++    public PlayerItemFrameChangeEvent(@NotNull Player player, @NotNull ItemFrame itemFrame,
++                                      @NotNull ItemStack itemStack, @NotNull ItemFrameChangeAction action) {
++        super(player);
++        this.itemFrame = itemFrame;
++        this.itemStack = itemStack;
++        this.action = action;
++    }
++
++    /**
++     * Gets the {@link ItemFrame} involved in this event.
++     * @return the {@link ItemFrame}
++     */
++    @NotNull
++    public ItemFrame getItemFrame() {
++        return itemFrame;
++    }
++
++    /**
++     * Gets the {@link ItemStack} involved in this event.
++     * This is the item being added or removed from the {@link ItemFrame}.
++     * @return the {@link ItemStack} being added or removed
++     */
++    @NotNull
++    public ItemStack getItemStack() {
++        return itemStack;
++    }
++
++    /**
++     * Sets the {@link ItemStack} that this {@link ItemFrame} holds.
++     * @param itemStack {@link ItemFrame} item
++     */
++    public void setItemStack(@NotNull ItemStack itemStack) {
++        this.itemStack = itemStack;
```

Comment:
Should probably have a null check here

Reply:
Not sure but if null is passed (ignoring the NotNull) the item resulting in NMS when convert is EMPTY.

Reply:
Then that should be documented and the nullability annotation changed

Reply:
For the sake of usability, I've made it nullable and then if its null just set the itemstack to air - this mimics the default behaviour if it were to be passed as null whilst making sure the `getItemStack` method never returns null.